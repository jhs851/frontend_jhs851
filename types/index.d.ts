export interface User {
  id: number
  name: string
  avatar: string
}

export interface Room {
  id: number
  user_id: number
  local_name: string
  unread_message_count: number
  last_message_id?: number
}

export interface RelatedRoom extends Room {
  user: User
  last_message?: Message
}

export interface Message {
  id: number
  user_id: number
  room_id: number
  message?: string
  asset?: string
  created_at: string
}

export interface RelatedMessage extends Message {
  user: User
  room: Room
}

export interface VirtualMessage extends Omit<Message, 'id'> {
  virtual_id: string
}

export interface Asset {
  id: number
  path: string
}
