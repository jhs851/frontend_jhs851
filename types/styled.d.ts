import 'styled-components'
import { Colors } from 'types/theme'

declare module 'styled-components' {
  export interface DefaultTheme extends Colors {}
}
