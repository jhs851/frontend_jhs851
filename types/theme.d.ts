export interface CommonColors {
  readonly black: string
  readonly white: string
  readonly grey1: string
  readonly coolGrey: string
  readonly paleGrey: string
  readonly battleshipGrey: string
  readonly charcoalGrey: string
  readonly charcoalGrey2: string
  readonly paleLilac: string
  readonly foreground: string
  readonly background: string
  readonly black10: string
}

export interface AccentColors {
  readonly primary: string
  readonly secondary: string
  readonly error: string
}

export interface Colors extends CommonColors, AccentColors {}

export type Theme = {
  light: Colors
  dark: Colors
}

export interface ThemeContextState {
  dark: boolean
  toggleTheme: () => void
  colors: Colors
}
