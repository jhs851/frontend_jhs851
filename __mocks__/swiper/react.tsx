import React from 'react'

const mock: React.FC = ({ children }) => <div>{children}</div>

export default mock
