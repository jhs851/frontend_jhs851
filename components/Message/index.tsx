import React, { useCallback, useEffect, useState } from 'react'
import Text from 'components/Text'
import { VirtualMessage as VMessage, RelatedMessage as RMessage } from 'types'
import {
  AssetContainer,
  AssetWrapper,
  CancelButton,
  RetryButton,
  FailedContainer,
  LoadingIcon,
  MessageContainer,
  Progress,
  ProgressContainer,
  StyledMessage,
  TimeText,
  RemoveButton,
} from './styled'
import { useTheme } from 'contexts/ThemeProvider'
import Image from 'components/Image'
import API from 'utils/API'
import { MdRefresh, MdClose } from 'react-icons/md'

type P = (RMessage | VMessage) & {
  isLast: boolean
  time: string
  onRemove: (message: VMessage) => void
}

const Message: React.FC<P> = props => {
  const { user_id, message, isLast, asset, time } = props
  const { colors } = useTheme()
  const isVirtualMessage = (
    message: RMessage | VMessage,
  ): message is VMessage => (message as VMessage).virtual_id !== undefined
  const [loading, setLoading] = useState<boolean>(isVirtualMessage(props))
  const [failed, setFailed] = useState<boolean>(false)
  const [width, setWidth] = useState<number>(0)
  const me = user_id === 1
  const store = useCallback(() => {
    if (isVirtualMessage(props)) {
      API.createMessage(props)
        .then(() => setLoading(false))
        .catch(() => setFailed(true))
    }
  }, [props])
  const onCancel = () => setFailed(true)
  const onRetry = () => {
    setFailed(false)
    setLoading(true)
    store()
  }
  const onRemove = () => {
    if (isVirtualMessage(props)) {
      props.onRemove(props)
    }
  }

  useEffect(() => {
    setWidth(95)
    store()
  }, [props, store])

  return (
    <MessageContainer me={me}>
      {!!message && loading && <LoadingIcon color={colors.coolGrey} />}

      {failed ? (
        <FailedContainer>
          <RetryButton onClick={onRetry}>
            <MdRefresh color={colors.charcoalGrey2} />
          </RetryButton>
          <RemoveButton onClick={onRemove}>
            <MdClose color={colors.error} />
          </RemoveButton>
        </FailedContainer>
      ) : (
        isLast &&
        (message || (asset && !loading)) && (
          <TimeText me={me} size={12} letterSpacing={0}>
            {time}
          </TimeText>
        )
      )}

      {message ? (
        <StyledMessage me={me}>
          {message
            .split('\n')
            .filter(value => value)
            .map((value, i) => (
              <Text
                key={i}
                size={14}
                weight={600}
                color={me ? 'white' : 'charcoalGrey2'}
              >
                {value}
              </Text>
            ))}
        </StyledMessage>
      ) : (
        !!asset && (
          <AssetContainer>
            <AssetWrapper>
              {loading && (
                <CancelButton onClick={onCancel}>
                  <Image
                    src="/icons/close.svg"
                    width={16}
                    height={16}
                    alt="취소"
                  />
                </CancelButton>
              )}

              <Image src={asset} width={200} height={200} alt="이미지" />
            </AssetWrapper>

            {loading && (
              <ProgressContainer>
                <Progress width={width} />
              </ProgressContainer>
            )}
          </AssetContainer>
        )
      )}
    </MessageContainer>
  )
}

export default Message
