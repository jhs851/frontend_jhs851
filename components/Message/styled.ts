import styled from 'styled-components'
import Text from 'components/Text'
import { HiOutlinePaperAirplane } from 'react-icons/hi'
import { flexCenter } from 'styles/flex'
import { motion } from 'framer-motion'

export const MessageContainer = styled.article<{ me: boolean }>`
  margin-top: 10px;
  display: flex;
  justify-content: ${({ me }) => (me ? 'flex-end' : 'flex-start')};
  align-items: flex-end;
`

export const StyledMessage = styled(motion.div)<{ me: boolean }>`
  padding: 12px;
  border-radius: 12px;
  box-shadow: 0 2px 4px 0 ${({ theme }) => theme.black10};
  word-break: break-word;
  background-color: ${({ theme, me }) =>
    me ? theme.primary : theme.background};
`

export const FailedContainer = styled.div`
  display: flex;
  margin: 0 4px;
  border-radius: 4px;
  overflow: hidden;
`

export const RetryButton = styled.button`
  width: 20px;
  height: 20px;
  background-color: ${({ theme }) => theme.background};
  border-right: 1px solid ${({ theme }) => theme.paleLilac};
  ${flexCenter};
`

export const RemoveButton = styled.button`
  width: 20px;
  height: 20px;
  background-color: ${({ theme }) => theme.background};
  ${flexCenter};
`

export const TimeText = styled(Text)<{ me: boolean }>`
  margin: 0 4px;
  opacity: 0.4;
  order: ${({ me }) => (me ? 0 : 1)};
`

export const AssetContainer = styled(motion.div)``

export const AssetWrapper = styled.div`
  position: relative;
  width: 200px;
  height: 200px;
  border-radius: 12px;
  overflow: hidden;
  box-shadow: 0 2px 4px 0 ${({ theme }) => theme.black10};
  background-color: ${({ theme }) => theme.charcoalGrey};
`

export const CancelButton = styled.button`
  position: absolute;
  width: 40px;
  height: 40px;
  opacity: 0.8;
  background-color: ${({ theme }) => theme.black};
  border-radius: 50%;
  left: 50%;
  top: 50%;
  margin-top: -20px;
  margin-left: -20px;
  z-index: 1030;
  cursor: pointer;
  transform: scale(1.2);
  ${flexCenter};
`

export const ProgressContainer = styled.div`
  height: 6px;
  border-radius: 3px;
  margin-top: 3px;
  background-color: ${({ theme }) => theme.grey1};
  overflow: hidden;
`

export const Progress = styled.div<{ width: number }>`
  width: ${({ width }) => width}%;
  height: 100%;
  transition: width 300ms linear;
  background-color: ${({ theme }) => theme.primary};
`

export const LoadingIcon = styled(HiOutlinePaperAirplane)`
  transform: rotate(-90deg) translateX(2px);
`
