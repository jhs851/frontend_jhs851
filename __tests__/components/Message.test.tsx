import { render } from '@testing-library/react'
import Message from 'components/Message'
import {
  messages as mockMessages,
  users as mockUser,
  rooms as mockRooms,
} from 'utils/mocks'
import { Message as IMessage, Room as IRoom, User as IUser } from 'types'

describe('Message 컴포넌트', () => {
  const message: IMessage = mockMessages[0]
  const user: IUser =
    mockUser.find(user => user.id === message.user_id) || mockUser[0]
  const room: IRoom =
    mockRooms.find(room => room.id === message.room_id) || mockRooms[0]
  const onRemove = jest.fn()
  const rendered = (isLast: boolean, time: string) =>
    render(
      <Message
        {...message}
        user={user}
        room={room}
        isLast={isLast}
        time={time}
        onRemove={onRemove}
      />,
    )

  it('렌더링이 올바르게 된다', () => {
    expect(rendered(true, '00:00').container).toMatchSnapshot()
  })
})
