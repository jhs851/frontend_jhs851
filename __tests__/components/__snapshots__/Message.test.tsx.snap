// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message 컴포넌트 렌더링이 올바르게 된다 1`] = `
.c1 {
  font-family: AppleSDGothicNeo,sans-serif;
  font-size: 12px;
  font-weight: 500;
  -webkit-letter-spacing: -0.1px;
  -moz-letter-spacing: -0.1px;
  -ms-letter-spacing: -0.1px;
  letter-spacing: -0.1px;
}

.c4 {
  font-family: AppleSDGothicNeo,sans-serif;
  font-size: 14px;
  font-weight: 600;
  -webkit-letter-spacing: -0.1px;
  -moz-letter-spacing: -0.1px;
  -ms-letter-spacing: -0.1px;
  letter-spacing: -0.1px;
}

.c0 {
  margin-top: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c3 {
  padding: 12px;
  border-radius: 12px;
  box-shadow: 0 2px 4px 0;
  word-break: break-word;
}

.c2 {
  margin: 0 4px;
  opacity: 0.4;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

<div>
  <article
    class="c0"
    style="opacity: 0; transform: translateX(30px) translateZ(0);"
  >
    <p
      class="c1 c2"
      letter-spacing="0"
      size="12"
    >
      00:00
    </p>
    <div
      class="c3"
    >
      <p
        class="c4"
        color="charcoalGrey2"
        size="14"
      >
        출근했니?
      </p>
    </div>
  </article>
</div>
`;
